src/main/java/org/matsim/run/rebalancing/WaitingPointsBasedRebalancingStrategy.java

```diff
@@ -0,0 +1,117 @@
+package org.matsim.run.rebalancing;
+
+import com.google.common.base.Preconditions;
+import org.apache.commons.csv.CSVFormat;
+import org.apache.commons.csv.CSVParser;
+import org.apache.commons.csv.CSVRecord;
+import org.apache.commons.lang.mutable.MutableInt;
+import org.matsim.api.core.v01.Id;
+import org.matsim.api.core.v01.network.Link;
+import org.matsim.api.core.v01.network.Network;
+import org.matsim.contrib.drt.optimizer.rebalancing.RebalancingParams;
+import org.matsim.contrib.drt.optimizer.rebalancing.RebalancingStrategy;
+import org.matsim.contrib.dvrp.fleet.DvrpVehicle;
+import org.matsim.contrib.dvrp.fleet.Fleet;
+import org.matsim.contrib.dvrp.schedule.Schedule;
+import org.matsim.contrib.dvrp.schedule.Schedules;
+import org.matsim.contrib.dvrp.schedule.StayTask;
+import org.matsim.core.utils.geometry.CoordUtils;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Stream;
+
+/**
+ * @author Chengqi Lu
+ */
+public class WaitingPointsBasedRebalancingStrategy implements RebalancingStrategy {
+
+	private final Network network;
+	private final RebalancingParams params;
+	private final Fleet fleet;
+	private final Map<Id<Link>, Integer> waitingPointsCapcityMap = new HashMap<>();
+
+	public WaitingPointsBasedRebalancingStrategy(Network network, String waitingPointsPath, RebalancingParams params, Fleet fleet) throws IOException {
+		this.network = network;
+		this.params = params;
+		this.fleet = fleet;
+		initialize(waitingPointsPath);
+	}
+
+	private void initialize(String waitingPointsPath) throws IOException {
+		// read tsv file: link_id capacity
+		try (CSVParser parser = new CSVParser(Files.newBufferedReader(Path.of(waitingPointsPath), StandardCharsets.UTF_8),
+			CSVFormat.TDF.builder().setHeader().setSkipHeaderRecord(true).build())) {
+			for (CSVRecord record : parser) {
+				Link waitingPointLink = network.getLinks().get(Id.createLinkId(record.get("link_id")));
+				Integer capacity = Integer.parseInt(record.get("capacity"));
+				waitingPointsCapcityMap.put(waitingPointLink.getId(), capacity);
+			}
+		}
+	}
+
+	@Override
+	public List<Relocation> calcRelocations(Stream<? extends DvrpVehicle> rebalancableVehicles, double time) {
+		List<Relocation> relocations = new ArrayList<>();
+		List<? extends DvrpVehicle> idleVehicles = rebalancableVehicles.filter(v -> v.getServiceEndTime() > time + params.minServiceTime).toList();
```

Comment:
is the order of the list random? Should it be? Maybe not for our small-fleet-size cases in Kelheim, but for other, large-fleet scenarios it might matter.
Because if for example the list is ordered by the vehicle Ids, it would always be the same vehicle(s) that have to clear a capacity-exceeded waiting point or that get to fill up the closes waiting point etc.

Reply:
Good point. I can add a shuffle operation to avoid this problem. 

But it needs to be pointed out that it doesn't guarantee the results will get better. It's just 2 different heuristic solutions, and sometimes one is better, and sometimes the other is better.

Reply:
That's true!